### DISCUSSION TITLE
BitrateLimit in Video Encoder Configuration Test

### DISCUSSION THREAD

**Discussion by JayDommaschk** [onvif/testspecs#79] [Q&A] [answered]
Hello,

in the video encoder configuration test [(specification)](https://www.onvif.org/wp-content/uploads/2021/06/ONVIF_Media_Configuration_Device_Test_Specification_20.12.pdf) it is stated that SetVideoEncoderConfiguration is being run with BitrateLimit = "64000" while every other value is taken from the response to GetVideoEncoderConfigurationOptions.

Wouldn't it be better to take the BitrateLimit from GetVideoEncoderConfigurationOptionsResponse, too, if it exists? (since it is in the Extension part, which is optional, the device might not include the BitrateRange part in its response) And as a fallback value if it doesn't exist or even if you insist to use a hardcoded value,  I think 64000 is a bit high considering that its unit is kpbs.

Usually I would check the parameter ranges of what I receive in SetVideoEncoderConfiguration and return an error if a value is outside the range, but because of this I cannot check this range lest I fail the conformance test. (luckily the Test Tool doesn't seem to care whether the BitrateLimit value is actually applied, though, but this is not a satisfying way of dealing with this)

**Comment by madhurao68** [comment, marked as answer]
Hello,
There is a requirement in ONVIF Media Service Specification, Section 5.5.5 SetVideoEncoderConfiguration:
"A device shall adapt an out of range BitrateLimit instead of returning a fault."

So, the client may send any values, and this is also applicable to the ONVIF Device Test Tool. Device shall not return the fault but adapt BitrateLimit to the supported one. Hope this helps!

**Reply by JayDommaschk** [reply]
Thank you. I indeed missed this. But then I would like to ask: What does "adapt" mean here? Should the device just internally set the highest possible BitrateLimit or would it actually have to save and return the out-of-range-value?

**Reply by baldvin137** [reply]
If I remember correctly (this goes back ten years or so) the intention was just to make sure that the device adjusts it internally (what you first state) so that a client does not have guess. I'm not sure if there is any requirement to save and return the out-of-range-value. The best way to find out is to read the test specification and see if there is any test that has this as an approval criteria. The test specifications are the normative interpretation of the network specifications.

**Reply by JayDommaschk** [reply]
I see. Thanks a lot.

At the very least, test 4.2.6 in the Media Test Specification (the one I linked in my start post) seems to be successful even if I ignore the out-of-range-value altogether (without returning an error). But I also tested with setting it to the highest bitrate and this was fine, too, so I will go for this and see if this causes a problem somewhere else.

So all my questions are answered. Thanks a lot to everyone and have a nice day.
